keyboards/bastardkb/charybdis/charybdis.c

```diff
@@ -0,0 +1,245 @@
+#endif
+        break;
+    }
+
+/* If Mousekeys is disabled, then use handle the mouse button
+ * keycodes.  This makes things simpler, and allows usage of
+ * the keycodes in a consistent manner.  But only do this if
+ * Mousekeys is not enable, so it's not handled twice.
+ */
+#ifndef MOUSEKEY_ENABLE
+    if (IS_MOUSEKEY_BUTTON(keycode)) {
+        report_mouse_t currentReport = pointing_device_get_report();
+        if (record->event.pressed) {
+            currentReport.buttons |= 1 << (keycode - KC_MS_BTN1);
+        } else {
+            currentReport.buttons &= ~(1 << (keycode - KC_MS_BTN1));
+        }
+        pointing_device_set_report(currentReport);
+        pointing_device_send();
+    }
+#endif
+
+    return true;
+}
+
+// Hardware Setup
+void keyboard_pre_init_kb(void) {
+    //debug_enable  = true;
+    //debug_matrix  = true;
+    //debug_mouse   = true;
+    scroll_inertia.x = 0;
+    scroll_inertia.y = 0;
+    // debug_encoder = true;
+    
+    keyboard_pre_init_user();
+}
+
+void pointing_device_init(void) {
+    pmw_spi_init();
+    pmw_set_cpi(dpi_array[keyboard_config.dpi_config]);
+}
+
+static bool has_report_changed(report_mouse_t new, report_mouse_t old) { return (new.buttons != old.buttons) || (new.x&& new.x != old.x) || (new.y&& new.y != old.y) || (new.h&& new.h != old.h) || (new.v&& new.v != old.v); }
+
+void pointing_device_task(void) {
+    report_mouse_t mouse_report = pointing_device_get_report();
+    process_mouse(&mouse_report);
+    
+    if (is_drag_scroll) {
+#ifdef CHARYBDIS_DRAGSCROLL_INVERT
+        // Invert vertical scroll direction
+        scroll_inertia.y += -mouse_report.y;
+#else
+        scroll_inertia.y += mouse_report.y;
+#endif
+        // we only want to trigger scrolling once in 12 scrolls, for slower scroll
+        if(scroll_inertia.y > CHARYBDIS_TRACKBALL_SPEED_DIVIDER || scroll_inertia.y < -CHARYBDIS_TRACKBALL_SPEED_DIVIDER){
+            mouse_report.v = constrain(scroll_inertia.y / CHARYBDIS_TRACKBALL_SPEED_DIVIDER * 2, 
+                                -CHARYBDIS_TRACKBALL_SPEED_DIVIDER/2, CHARYBDIS_TRACKBALL_SPEED_DIVIDER/2);
+            if(debug_mouse) uprintf("Scroll] V: %d // %d\n", mouse_report.v, scroll_inertia.y);
+            scroll_inertia.y = 0;
+        }
+
+        mouse_report.x = 0;
+        mouse_report.y = 0;
+    }
+
+    pointing_device_set_report(mouse_report);
+    pointing_device_send();
+}
+
+void eeconfig_init_kb(void) {
+    keyboard_config.dpi_config = CHARYBDIS_DPI_DEFAULT;
+    keyboard_config.sniper_config = CHARYBDIS_SNIPER_DEFAULT;
+    pmw_set_cpi(dpi_array[keyboard_config.dpi_config]);
+    eeconfig_update_kb(keyboard_config.raw);
+    eeconfig_init_user();
+}
+
+void matrix_init_kb(void) {
+    // is safe to just read DPI setting since matrix init
+    // comes before pointing device init.
+    keyboard_config.raw = eeconfig_read_kb();
+    if (keyboard_config.dpi_config > DPI_OPTION_SIZE) { eeconfig_init_kb(); }
```

Comment:
Should this be `>=`?

```c
if (keyboard_config.dpi_config >= DPI_OPTION_SIZE) {
  eeconfig_init_kb();
}
```

Reply:
Should we also guard `.sniper_config`?

```c
if (keyboard_config.dpi_config >= DPI_OPTION_SIZE ||
    keyboard_config.sniper_config >= SNIPER_OPTION_SIZE) {
  eeconfig_init_kb();
}
```

Reply:
Arrays start at 0, so I think it's `>`.
Good catch on the `sniper_config`, added.